### ISSUE TITLE
Offset/alignment for MallocAllocator

### ISSUE THREAD

**Issue by qinsoon** [mmtk/mmtk-core#351]
Our MallocAllocator does not allow alignment larger than 16 bytes, and offset is ignored. This is mainly because malloc does not support alignment and offset. However, we should still support offset/alignment.

* Alignment: we can use aligned_alloc or posix_memalign. If the malloc library does not provide an aligned allocation, we need to do the alignment ourselves.
* Offset: we need to allocate slightly larger than the required size so we can return an offset result. Also we need to record the malloc result, and when we free an object, we need to find the malloc result for the object (we can do something similar to find base reference for an internal reference).

**Comment by YiluoWei** [comment]
When we free an object, we need to first know whether it is allocated with an offset, and if it is, we need to find the malloc result for the object. 

My question is who, the VM or mmtk-core, should record whether an object is allocated with an offset, and how. 

I can see now in mmtk-core, we ask the binding through `VM::VMObjectModel::object_start_ref` to get the pointer to be freed.
Also, alloc is called from the VM, and the VM may have an allocation fastpath that may by-pass mmtk-core.

**Comment by qinsoon** [comment]
> When we free an object, we need to first know whether it is allocated with an offset, and if it is, we need to find the malloc result for the object.
> 
> My question is who, the VM or mmtk-core, should record whether an object is allocated with an offset, and how.

Definitely MMTk. We could use another side metadata to record the malloc result.

> I can see now in mmtk-core, we ask the binding through `VM::VMObjectModel::object_start_ref` to get the pointer to be freed.
> Also, alloc is called from the VM, and the VM may have an allocation fastpath that may by-pass mmtk-core.

There are 3 things here: the object reference, the cell address we return in an alloc call, and the malloc result. Currently the cell address always equals the malloc result, and `object_start_ref()` will give us the cell address. So we can use it to free objects. But with offset and alignment, the cell address may be different than the malloc result. I guess we have to somehow record the malloc result. We could use another side metadata.

**Comment by YiluoWei** [comment]
If we use side metadata, instead of another side metadata, we should probably use 2 bits for allocation side metadata, something like 
01 : there is an object allocated here
11 : there is an object allocated with an offset here

so that we only need load/store one kind of side metadata when we free/allocate an object. 
but I also notice that `sweep_chunk_mark_on_side` requires the alloc bit to be excat 1 bit.


For libc, jemalloc and mimalloc, we can use `posix_memalign()` and the result can be freed directly using `free()`. For hoard, we may need to do the alignment ourselves and record the malloc result.

I see a common method to record the malloc result. 
If we want to malloc(`size` , `alignment`, `offset`)
we malloc extra `sizeof(*void)` bytes, and also extra `alignment` bytes, and store the malloc result in front of the cell address.

```
-----------------  <- malloc result
| empty
-----------------  <- (cell_address - `sizeof(*void)`) 
| malloc result
-----------------  <- cell_address
| something
-----------------  <- cell_address + offset, which is aligned to the given alignment
| ........
-----------------  <- cell_address + size
| empty
-----------------  <- malloc_result + size + `sizeof(*void) + `alignment`
```

**Comment by YiluoWei** [comment]
If we use side metadata, instead of another side metadata, we should probably use 2 bits for allocation side metadata, something like 
01 : there is an object allocated here
11 : there is an object allocated with an offset here

so that we only need load/store one kind of side metadata when we free/allocate an object. 
but I also notice that `sweep_chunk_mark_on_side` requires the alloc bit to be excat 1 bit.


For libc, jemalloc and mimalloc, we can use `posix_memalign()` and the result can be freed directly using `free()`. For hoard, we may need to do the alignment ourselves and record the malloc result.

I see a common method to record the malloc result. 
If we want to malloc(`size` , `alignment`, `offset`)
we malloc extra `sizeof(*void)` bytes, and also extra `alignment` bytes, and store the malloc result in front of the cell address.

```
-----------------  <- malloc result
| empty
-----------------  <- (cell_address - `sizeof(*void)`) 
| malloc result
-----------------  <- cell_address
| something
-----------------  <- cell_address + offset, which is aligned to the given alignment
| ........
-----------------  <- cell_address + size
| empty
-----------------  <- malloc_result + size + `sizeof(*void) + `alignment`
```

**Comment by YiluoWei** [comment]
>> 11 : there is an object allocated with an offset here

> But we don't know the offset right? How does this give us the proper object reference from the malloc result?

When we free an object, we have the object reference, but we actually need the malloc result. If we find the metadata to be `11`, we will know that we need to find the malloc result, which, as descriped above, shoule be stored at `(object_ref - sizeof(void*))`.

> As Yi and I discussed on the relevant PR here, we don't need posix_memalign() for mimalloc, jemalloc, and Hoard.

Do you mean that we use `aligned_alloc()` instead of `posix_memalign()`?
from https://linux.die.net/man/3/posix_memalign 
```
The function aligned_alloc() is the same as memalign(), except for the added restriction that size should be a multiple of alignment.
```

**Comment by k-sareen** [comment]
> When we free an object, we have the object reference, but we actually need the malloc result. If we find the metadata to be 11, we will know that we need to find the malloc result, which, as descriped above, shoule be stored at (object_ref - sizeof(void*))

Ah that makes sense.

> Do you mean that we use aligned_alloc() instead of posix_memalign()? From https://linux.die.net/man/3/posix_memalign

I'm aware that `aligned_alloc()` will internally call `posix_memalign()`, although this is malloc-library implementation dependant (I think libc `alligned_alloc()` will call `posix_memalign()`). I guess more precisely what I meant was we should use the correct alloc function for each library. Mimalloc, for example, will not call `posix_memalign()` internally. Though mimalloc has the `aligned_calloc()` function as well which does not require an extra `memset()` operation which would be preferable over `aligned_alloc()`.

**Comment by qinsoon** [comment]
Also it is possible to deal with the alignment in mmtk-core rather than using `aligned_alloc()` from the library. The following is how Java MMTk does for a freelist allocator. We have the same `get_maximum_aligned_size()` and `align_allocation()` implemented.
* Use `getMaximumAlignedSize()` to get a size before calling malloc/calloc: https://github.com/JikesRVM/JikesRVM/blob/38b21f5a663016dbf43771cb2d231f74db0a01c6/MMTk/src/org/mmtk/utility/alloc/SegregatedFreeList.java#L74
* Use `alignAllocation` for alignment and offset: https://github.com/JikesRVM/JikesRVM/blob/38b21f5a663016dbf43771cb2d231f74db0a01c6/MMTk/src/org/mmtk/utility/alloc/SegregatedFreeList.java#L81

**Comment by k-sareen** [comment]
Interesting -- this would make life easier to a certain extent as we would not have to do the `memset()` after calling `aligned_alloc()`. I think pragmatically, what you have outlined above is probably a better solution.

**Comment by k-sareen** [comment]
Closed with #475
